paddle/fluid/pir/transforms/general/transfer_layout_pass.cc

```diff
@@ -0,0 +1,777 @@
+    float ret = 0.0f;
+    while (next_arc < adjs[src].size()) {
+      auto e_ind = adjs[src][next_arc];
+      auto& e = edges[e_ind];
+      next_arc++;
+      auto next_node = e.dst;
+      if (heights[next_node] == heights[src] + 1) {
+        auto left_capacity = e.capacity - e.flow;
+        auto update_flow = std::min(cf - ret, left_capacity);
+        auto f = FindBlockingFlow(next_node, update_flow);
+        if (f > 0) {
+          e.flow += f;
+          auto reverse_e_ind = e_ind ^ 1;
+          auto& reverse_e = edges[reverse_e_ind];
+          reverse_e.flow -= f;
+          ret += f;
+
+          if (abs(ret - cf) < 1e-9) {
+            return ret;
+          }
+        }
+      }
+    }
+
+    if (ret == 0) {
+      heights[src] = 0;
+    }
+
+    return ret;
+  }
+
+  float max_flow() {
+    VLOG(10)
+        << "--------------------[max flow start]---------------------------";
+    float total_flow = 0.0f;
+    while (ConstructLevelGraph()) {
+      for (auto& [node, nexts] : adjs) {
+        cur_arcs[node] = 0;
+      }
+      while (auto f = FindBlockingFlow(src_node(), INF)) {
+        total_flow += f;
+      }
+    }
+    VLOG(10) << "--------------------[max flow end]---------------------------";
+    return total_flow;
+  }
+
+  std::tuple<std::unordered_set<Node>, std::vector<Edge>> min_cut() {  // NOLINT
+    max_flow();
+    // from src_node get its reachable nodes and call them S
+    // other nodes are in T
+    // collect edges between S and T
+    std::unordered_set<Node> src_set;
+    std::queue<Node> q;
+    q.push(src_node());
+    while (!q.empty()) {
+      auto n = q.front();
+      q.pop();
+      VLOG(10) << "bfs access: " << n;
+      if (src_set.count(n) > 0) continue;
+      src_set.insert(n);
+      VLOG(10) << "bfs insert " << n << " " << src_set.size();
+      for (auto& ind : adjs[n]) {
+        VLOG(10) << "bfs edge: " << edges[ind] << " c:" << edges[ind].capacity
+                 << " f:" << edges[ind].flow;
+        if (edges[ind].capacity > edges[ind].flow) {
+          VLOG(10) << "bfs add: " << edges[ind].dst;
+          q.push(edges[ind].dst);
+        }
+      }
+    }
+
+    VLOG(10) << "src_set.size()=" << src_set.size();
+
+    std::vector<Edge> cut;
+    for (const auto& e : edges) {
+      if (!e.real) continue;
+      auto& src = e.src;
+      auto& dst = e.dst;
+      bool src_cond = (src_set.count(src) > 0);
+      bool dst_cond = (src_set.count(dst) > 0);
+      if (src_cond == dst_cond) {
+        continue;
+      }
+      VLOG(10) << "cut " << src << "(" << src_cond << ")"
+               << " " << dst << "(" << dst_cond << ")";
+      cut.push_back(e);
+    }
+
+    VLOG(10) << "cut set.size()=" << cut.size();
+    VLOG(10) << "-----------------------------------------------";
+
+    return {src_set, cut};
+  }
+};
+
+using Edge = FlowGraph::Edge;
+
+class TransferLayoutPass : public pir::Pass {
+ public:
+  TransferLayoutPass() : pir::Pass("transfer_layout_pass", 2) {}
```

Comment:
```suggestion
  TransferLayoutPass() : pir::Pass("transfer_layout_pass", 3 or 4) {}
```
不建议引入config.EnableTransferLayout这个api来控制。opt level可以设为3或者4，然后通过config.SetOptimizationLevel api来控制就可以了。

Reply:
如果transfer_layout_pass稳定的话，opt level直接设为2，也不是不可以

Reply:
好的，我先去掉，目前还是不够稳定，设为3吧。

Reply:
设为了4。